lib/hammer_cli_foreman/api/authenticator.rb

```diff
@@ -0,0 +1,96 @@
+module HammerCLIForeman
+  module Api
+    class Authenticator
+      attr_accessor :auth_type, :uri, :settings
+      def initialize(auth_type, uri, settings)
+        @auth_type = auth_type
+        @uri = uri
+        @settings = settings
+      end
+
+      def fetch
+        if ssl_cert_authentication? && !use_basic_auth?
+          void_auth
+        elsif auth_type == AUTH_TYPES[:basic_auth]
+          basic_auth
+        elsif auth_type == AUTH_TYPES[:oauth_without_2fa]
+          oauth_without_2fa
+        elsif auth_type == AUTH_TYPES[:oauth_with_2fa]
+          ouath_with_2fa
+        end
+      end
+
+      private
+
+      def void_auth
+        VoidAuth.new(_('Using certificate authentication.'))
+      end
+
+      def basic_auth
+        if HammerCLIForeman::Sessions.enabled?
+          authenticator = InteractiveBasicAuth.new(
+            settings.get(:_params, :username) || ENV['FOREMAN_USERNAME'],
+            settings.get(:_params, :password) || ENV['FOREMAN_PASSWORD']
+          )
+          SessionAuthenticatorWrapper.new(authenticator, uri, auth_type)
+        else
+          username = settings.get(:_params, :username) || ENV['FOREMAN_USERNAME'] || settings.get(:foreman, :username)
+          password = settings.get(:_params, :password) || ENV['FOREMAN_PASSWORD']
+          if password.nil? && (username == settings.get(:foreman, :username))
+            password = settings.get(:foreman, :password)
+          end
+          InteractiveBasicAuth.new(username, password)
+        end
+      end
+
+      def oauth_without_2fa
+        if HammerCLIForeman::Sessions.enabled?
+          authenticator = Oauth::InteractiveKeycloakAuth.new(
+            settings.get(:_params, :url) || ENV['KEYCLOAK_URL'],
+            settings.get(:_params, :realm) || ENV['KEYCLOAK_REALM'],
+            settings.get(:_params, :client_id) || ENV['KEYCLOAK_CLIENT_ID'],
+            settings.get(:_params, :username) || ENV['KEYCLOAK_USERNAME'],
+            settings.get(:_params, :password) || ENV['KEYCLOAK_PASSWORD']
+          )
+          SessionAuthenticatorWrapper.new(authenticator, uri, auth_type)
+        else
```

Comment:
This part of the condition will not work (dead code) for oauth with/out because of https://github.com/theforeman/hammer-cli-foreman/blob/f0c3cecb34dc897dc4931a3b29f3a5cb7a118f74/lib/hammer_cli_foreman/api/connection.rb#L53

It's always `basic_auth` if sessions are disabled. Please, correct me if I'm wrong.

Reply:
Correct, it's always `basic_auth` if sessions are disabled. The pointed code above is dead code.